MetaMask/metamask-extension
# fix: disable UI buttons for accounts that cannot sign

## **Description**

The buttons for "Buy & Sell," "Send," "Swap," and "Bridge" should be disabled in the UI if the selected account doesn't support the method `eth_signTransaction`

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23178?quickstart=1)

## **Related issues**

Fixes: [Disable UI buttons for accounts that cannot sign transactions](https://app.zenhub.com/workspaces/metamask-accounts-team-v2-64c91cbeaa9d1c00126621fd/issues/gh/metamask/accounts-planning/271)

## **Manual testing steps**

1. Go to the [SSK](https://metamask.github.io/snap-simple-keyring/latest/) and create an account
2. Update the account using this JSON, replacing the address and ID fields with the new account information (this update step removes `eth_signTransaction` from the account's methods)
```json
{
    "address": "you new account address",
    "id": "your new account ID",
    "options": {},
    "methods": [
        "personal_sign",
        "eth_sign",
        "eth_signTypedData_v1",
        "eth_signTypedData_v3",
        "eth_signTypedData_v4"
    ],
    "type": "eip155:eoa"
}
```
3. Open the extension client and hover over the buttons for "Buy & Sell," "Send," "Swap," and "Bridge"
4. You should NOT be able to click the "Buy & Sell", "Send," "Swap," and "Bridge" buttons and you should see a tooltip with the message "This account can't sign transactions."

## **Jest**

`jest eth-overview.test.js`
`jest eth-overview.test.js -t "Disabled buttons when an account cannot sign transactions"`

## **Screenshots/Recordings**

### **Before**

Buttons for methods an account cannot support are enabled

<img width="453" alt="Screenshot 2024-02-26 at 11 59 31 AM" src="https://github.com/MetaMask/metamask-extension/assets/91970214/74c3e7fc-8a0d-4aea-984e-fe47ae7c5055">

<img width="362" alt="Screenshot 2024-02-26 at 12 02 45 PM" src="https://github.com/MetaMask/metamask-extension/assets/91970214/59af512b-d182-4279-916d-49ddce3784d3">

### **After**

Buttons for unsupported methods are now disabled with message

https://github.com/MetaMask/metamask-extension/assets/91970214/e12f2b6a-7fbc-4226-989a-1eb73213bd4f

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
